### PR TITLE
Initial CIF support

### PR DESCRIPTION
**Pull request by fxcoudert**
This is going to be, for now, work in progress. It does not compile. But I wanted to have first feedback on it, in particular if I understand correctly what are the files to modify.

The only function I have written for now is `CIFFormat::write` 😄

### CODE REVIEW

File: src/formats/CIF.cpp
Comment (Luthaf) [inline src/formats/CIF.cpp:379]: the closest we have is the "name" frame property, which you can get with `frame.get("name").as_string()` or something close.

File: src/formats/CIF.cpp
Comment (Luthaf) [inline src/formats/CIF.cpp:381]: Could this use `frame.cell.shape` instead to set the setting to orthorhombic? There is also the question of what to do with infinite cells.

File: src/formats/CIF.cpp
Comment (Luthaf) [inline src/formats/CIF.cpp:410]: no, everything is stored in cartesian format

File: src/formats/CIF.cpp
Comment (Luthaf) [inline src/formats/CIF.cpp:413]: not yet, but such occupancy could be stored as a new Atom property, read in by all format supporting it and checked when writing.

That's how we deal with PDB insertion code or ATOM/HETATM distinction for examle

**Comment by fxcoudert** [comment]
👍 
- First complete draft
- Compiles if you use `CXXFLAGS=-I/path/to/gemmi-0.4.0/include`

👎 
- No doc, no test
- Plenty of TODO's
- No build integration for gemmi

File: README.md
Comment (fxcoudert) [inline README.md:167]:
```suggestion
Chemfiles depends on multiple external libraries, which are distributed under [their
```

File: src/formats/CIF.cpp
Comment (fxcoudert) [reply src/formats/CIF.cpp:120]: Even if the cell has equal cell vectors and angles of 90°, in crystallographic terms as long as its space group is P1 it is in the triclinic cell setting.

Comment (fxcoudert) [reply src/formats/CIF.cpp:120]: Regarding infinite cells, other packages are facing the same question, and it's not clear what is the best answer (see for example https://gitlab.com/ase/ase/-/issues/286).

What Mercury does is add a fake cubic cell with 1 Å size, so fractional coordinates are equal to Cartesian:

```
_cell_length_a                   1
_cell_length_b                   1
_cell_length_c                   1
_cell_angle_alpha                90
_cell_angle_beta                 90
_cell_angle_gamma                90
```

Comment (Luthaf) [reply src/formats/CIF.cpp:120]: > Even if the cell has equal cell vectors and angles of 90°, in crystallographic terms as long as its space group is P1 it is in the triclinic cell setting.

Oh, ok that's good to know!

> What Mercury does is add a fake cubic cell with 1 Å size, so fractional coordinates are equal to Cartesian

That sounds good, I think the same convention is applied in PDB and potentially mmCIF files when no cell information is present. The issue would be when reading such file, we would like to create an infinite cell back, not a fake cubic cell. A potential criteria to distinguish fake 1A cubic cells from real ones could be the presence of atoms with coordinates of 4A or more. Although real 1A cubic cells should be quite uncommon anyway ^^

**Comment by Luthaf** [comment]
Sorry for the force push, I was in the wrong terminal 😐 I hope nothing was lost.

**Comment by fxcoudert** [comment]
Apparently now I can't pull and push my local changes (done in the train). Here they are:

```
commit 17914f428d03ff6930311c677ee19264898bdbb0
Author: FX Coudert <fxcoudert@gmail.com>
Date:   2020-07-23 14:46:55 +0200

    Add CIF tests

diff --git a/tests/formats/cif.cpp b/tests/formats/cif.cpp
new file mode 100644
index 00000000..2f884be5
--- /dev/null
+++ b/tests/formats/cif.cpp
@@ -0,0 +1,99 @@
+// Chemfiles, a modern library for chemistry file reading and writing
+// Copyright (C) Guillaume Fraux and contributors -- BSD license
+
+#include <streambuf>
+#include <fstream>
+
+#include "catch.hpp"
+#include "helpers.hpp"
+#include "chemfiles.hpp"
+using namespace chemfiles;
+
+TEST_CASE("Read files in CIF format") {
+    SECTION("Read single step") {
+        // This is how I imagine most people will resolve the conflict between
+        // CIF files and mmCIF files.
+        auto file = Trajectory("data/cif/1544173.cif", 'r', "CIF");
+        Frame frame = file.read();
+
+        CHECK(frame.size() == 100);
+	CHECK(frame.name == "1544173");
+
+        // Check reading positions
+        auto positions = frame.positions();
+        CHECK(approx_eq(positions[0], Vector3D(-0.428, 5.427, 11.536), 1e-3));
+        CHECK(approx_eq(positions[1], Vector3D(4.634, 11.178, 0.405), 1e-3));
+	CHECK(frame[0].name == "O1");
+	CHECK(frame[1].name == "O1");
+
+        // Check the unit cell
+        auto cell = frame.cell();
+        CHECK(approx_eq(cell.a(), 5.9170, 1e-4));
+        CHECK(approx_eq(cell.b(), 11.5030, 1e-4));
+        CHECK(approx_eq(cell.c(), 12.0635, 1e-4));
+        CHECK(approx_eq(cell.alpha(), 90., 1e-4));
+        CHECK(approx_eq(cell.beta(), 98.153, 1e-4));
+        CHECK(approx_eq(cell.gamma(), 90., 1e-4));
+    }
+
+    SECTION("Check nsteps") {
+        auto file = Trajectory("data/cif/1544173.cif", 'r', "CIF");
+        CHECK(file.nsteps() == 1);
+    }
+}
+
+TEST_CASE("Write CIF file") {
+    auto tmpfile = NamedTempPath(".cif");
+    const auto EXPECTED_CONTENT =
+    "# generated by Chemfiles\n"
+    "#\n"
+    "_cell.length_a 22.0\n"
+    "_cell.length_b 22.0\n"
+    "_cell.length_c 22.0\n"
+    "_cell.length_alpha 90.0\n"
+    "_cell.length_beta  90.0\n"
+    "_cell.length_gamma 90.0\n"
+    "#\n"
+    "loop_\n"
+    "_atom_site.group_PDB\n"
+    "_atom_site.id\n"
+    "_atom_site.type_symbol\n"
+    "_atom_site.label_atom_id\n"
+    "_atom_site.label_alt_id\n"
+    "_atom_site.label_comp_id\n"
+    "_atom_site.label_asym_id\n"
+    "_atom_site.label_seq_id\n"
+    "_atom_site.Cartn_x\n"
+    "_atom_site.Cartn_y\n"
+    "_atom_site.Cartn_z\n"
+    "_atom_site.pdbx_formal_charge\n"
+    "_atom_site.auth_asym_id\n"
+    "_atom_site.pdbx_PDB_model_num\n"
+    "HETATM 1     A  A    .   . .    .    1.000    2.000    3.000 0.0 . 1\n"
+    "ATOM   2     B  B    . foo ?    2    1.000    2.000    3.000 0.0 . 1\n"
+    "ATOM   3     C  C    . foo ?    2    1.000    2.000    3.000 0.0 . 1\n"
+    "HETATM 4     D  D    . bar G    ?    1.000    2.000    3.000 0.0 A 1\n"
+    "HETATM 5     A  A    .   . .    .    4.000    5.000    6.000 0.0 . 2\n"
+    "ATOM   6     B  B    . foo ?    2    1.000    2.000    3.000 0.0 . 2\n"
+    "ATOM   7     C  C    . foo ?    2    1.000    2.000    3.000 0.0 . 2\n"
+    "HETATM 8     D  D    . bar G    ?    1.000    2.000    3.000 0.0 A 2\n";
+
+
+    auto frame = Frame(UnitCell(22));
+    frame.add_atom(Atom("A"), {1, 2, 3});
+    frame.add_atom(Atom("B"), {1, 2, 3});
+    frame.add_atom(Atom("C"), {1, 2, 3});
+    frame.add_atom(Atom("D"), {1, 2, 3});
+
+    auto file = Trajectory(tmpfile, 'w');
+    file.write(frame);
+
+    frame.positions()[0] = {4.0, 5.0, 6.0};
+    file.write(frame);
+
+    file.close();
+    std::ifstream checking(tmpfile);
+    std::string content((std::istreambuf_iterator<char>(checking)),
+                         std::istreambuf_iterator<char>());
+    CHECK(EXPECTED_CONTENT == content);
+}
```

and

```
commit 13a3d03c69fbef5733caf41faa1bd4403535af39
Author: FX Coudert <fxcoudert@gmail.com>
Date:   2020-07-23 14:46:37 +0200

    Fix CIF output, handle infinite cells

diff --git a/src/formats/CIF.cpp b/src/formats/CIF.cpp
index 8f19fcac..bece85d0 100644
--- a/src/formats/CIF.cpp
+++ b/src/formats/CIF.cpp
@@ -115,17 +115,37 @@ void CIFFormat::read(Frame& frame) {
 }
 
 void CIFFormat::write(const Frame& frame) {
-    file_.print("data_model_{}\n", models_); // TODO: do models have a name in Chemfiles? If so, use it
+    auto name = frame.get("name");
+    if (name && name->kind() == Property::STRING) {
+        file_.print("data_{}\n", name->as_string());
+    } else {
+        file_.print("data_model_{}\n", models_);
+    }
     file_.print("_audit_creation_method         'generated by Chemfiles'\n");
     file_.print("_symmetry_cell_setting         'triclinic'\n");
     file_.print("_symmetry_space_group_name_H-M 'P 1'\n");
     file_.print("_space_group_IT_number         1\n");
-    file_.print("_cell.length_a {}\n", frame.cell().a());
-    file_.print("_cell.length_b {}\n", frame.cell().b());
-    file_.print("_cell.length_c {}\n", frame.cell().c());
-    file_.print("_cell.length_alpha {}\n", frame.cell().alpha());
-    file_.print("_cell.length_beta  {}\n", frame.cell().beta());
-    file_.print("_cell.length_gamma {}\n", frame.cell().gamma());
+
+    // We handle infinite cells like Mercury
+    Matrix3D invmat = Matrix3D::unit();
+    double a = 1, b = 1, c = 1;
+    double alpha = 90, beta = 90, gamma = 90;
+    if (frame.cell().shape() != UnitCell::INFINITE) {
+      invmat = frame.cell().matrix().invert();
+      a = frame.cell().a();
+      b = frame.cell().b();
+      c = frame.cell().c();
+      alpha = frame.cell().alpha();
+      beta = frame.cell().beta();
+      gamma = frame.cell().gamma();
+    }
+
+    file_.print("_cell.length_a {}\n", a);
+    file_.print("_cell.length_b {}\n", b);
+    file_.print("_cell.length_c {}\n", c);
+    file_.print("_cell.length_alpha {}\n", alpha);
+    file_.print("_cell.length_beta  {}\n", beta);
+    file_.print("_cell.length_gamma {}\n", gamma);
     file_.print("loop_\n");
     file_.print("  _symmetry_equiv_pos_as_xyz\n");
     file_.print("  '+x,+y,+z'\n");
@@ -144,16 +164,11 @@ void CIFFormat::write(const Frame& frame) {
     file_.print("_atom_site_Cartn_y\n");
     file_.print("_atom_site_Cartn_z\n");
 
-
     const auto& positions = frame.positions();
     for (size_t i = 0; i < frame.size(); ++i) {
         const auto& atom = frame[i];
-
-	// TODO: handle infinite cells
-        Matrix3D invmat = frame.cell().matrix().invert();
 	Vector3D fract = invmat * positions[i];
-
-        file_.print("{} {} 1.0 {: <4} {} {: >3} {} {: >4} {:8.3f} {:8.3f} {:8.3f} {} {} {}\n",
+        file_.print("{} {} 1.0 {:10.7f} {:10.7f} {:10.7f} {:8.5f} {:8.5f} {:8.5f}\n",
                 atom.name(), atom.type(), fract[0], fract[1], fract[2],
                 positions[i][0], positions[i][1], positions[i][2]
         );
```

With this I can successfully use cfiles to and from CIF files.

**Comment by fxcoudert** [comment]
@Luthaf feedback welcome on the remaining TODO's and the tests

**Comment by Luthaf** [comment]
There seems to be multiple memory leaks (see [valgrind](https://travis-ci.org/github/chemfiles/chemfiles/jobs/711927944) test on CI) coming from pegtl (AFAIK the parser generator used by Gemmi), from these functions and some others:

```
bool tao::pegtl::internal::duseltronik<gemmi::cif::rules::loop_value, (tao::pegtl::apply_mode)1, (tao::pegtl::rewind_mode)0, gemmi::cif::Action, gemmi::cif::Errors, (tao::pegtl::dusel_mode)2>::match<tao::pegtl::memory_input<(tao::pegtl::tracking_mode)0, tao::pegtl::ascii::eol::lf_crlf, std::string>, gemmi::cif::Document&>(tao::pegtl::memory_input<(tao::pegtl::tracking_mode)0, tao::pegtl::ascii::eol::lf_crlf, std::string>&, gemmi::cif::Document&) (duseltronik.hpp:87)

bool tao::pegtl::internal::rule_conjunction<gemmi::cif::rules::loop_value, gemmi::cif::rules::ws_or_eof, tao::pegtl::discard>::match<(tao::pegtl::apply_mode)1, (tao::pegtl::rewind_mode)0, gemmi::cif::Action, gemmi::cif::Errors, tao::pegtl::memory_input<(tao::pegtl::tracking_mode)0, tao::pegtl::ascii::eol::lf_crlf, std::string>, gemmi::cif::Document&>(tao::pegtl::memory_input<(tao::pegtl::tracking_mode)0, tao::pegtl::ascii::eol::lf_crlf, std::string>&, gemmi::cif::Document&) (rule_conjunction.hpp:51)
```

---

A few other leaks come from `chemfiles::CIFFormat::init_()`.

File: tests/formats/cif.cpp
Comment (Luthaf) [inline tests/formats/cif.cpp:3]: this should be wrapped in `#include <chemfiles.hpp> #idndef CHFL_DISABLE_GEMMI ... #endif`

**Comment by Luthaf** [comment]
One thing we will have to address is the fact that both wwwPDB and IUCr use `.cif` as the extension and very similar semantics of the STAR file. At the minimum, I think we should try to detect the presence of mmCIF tags in the CIF reader, and symmetry/space group tags in the mmCIF reader to send a warning and suggest using the other reader. Merging both reader could also be nice, but can happen later.

File: src/formats/CIF.cpp
Comment (Luthaf) [inline src/formats/CIF.cpp:103]: weight can be used as the atom mass `atom.set_mass(site.weight())`, with the caveat that chemfiles expects masses in atomic mass units. I don't know what's the standard in CIF files

for the occupancy, I would set it as a property `atom.set("occupancy", site.occ)` or `atom.set("site_occupancy", site.occ)`.

Atomic number can not be overridden, so it could go in a property too `atom.set("cif_atomic_number", site.atomic_number())`

File: src/formats/CIF.cpp
Comment (Luthaf) [inline src/formats/CIF.cpp:123]: seems like this was taken from mmCIF, which uses `model` as the trajectory step. Instead we could use `current_step_` here

File: src/formats/CIF.cpp
Comment (Luthaf) [inline src/formats/CIF.cpp:130]: Maybe add a link to Mercury/add more explanations of what does that mean?

File: src/formats/CIF.cpp
Comment (Luthaf) [inline src/formats/CIF.cpp:92]: This will not be able to read back a file written with the Mercury convention.

File: tests/formats/cif.cpp
Comment (Luthaf) [inline tests/formats/cif.cpp:15]: copy/paste scories =)

File: tests/formats/cif.cpp
Comment (Luthaf) [inline tests/formats/cif.cpp:16]: It looks like the CIF reader is able to read mmCIF files, but I guess it will ignore any kind of secondary structure information. We should send a warning in this case.

File: tests/formats/cif.cpp
Comment (Luthaf) [inline tests/formats/cif.cpp:48]: Could you add checks that atom names, properties & position are read as expected?

File: tests/formats/cif.cpp
Comment (Luthaf) [inline tests/formats/cif.cpp:61]: The tests are not using CAN.cif and SFE.cif, is this by design?

File: src/formats/CIF.cpp
Comment (fxcoudert) [reply src/formats/CIF.cpp:74]: Actually we want to throw here, the comment is wrong.

File: src/formats/CIF.cpp
Comment (fxcoudert) [reply src/formats/CIF.cpp:69]: Actually, one site can have multiple images, but it has always at least one. So we can simplify this logic…

File: src/formats/CIF.cpp
Comment (fxcoudert) [reply src/formats/CIF.cpp:103]: Only occupancy is native to CIF format, the other two are actually determined based on element name by Gemmi… so I decided to only keep the occupancy.

**Comment by fxcoudert** [comment]
I see valgrind complaining about:

```
Invalid read of size 1
  at 0x6B545C: void gemmi::split_element_and_charge<gemmi::SmallStructure::Site>(std::string const&, gemmi::SmallStructure::Site*) (small.hpp:77)
  by 0x6B1CB4: gemmi::make_small_structure_from_block(gemmi::cif::Block const&) (smcif.hpp:73)
  by 0x6A4AD5: chemfiles::CIFFormat::init_() (CIF.cpp:67)
```

which comes from some string operations in Gemmi's parser… not sure if real or a false positive, will try to report it to Gemmi if I can reproduce.

Also:

```
72,704 bytes in 1 blocks are still reachable in loss record 1 of 1
  at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
  by 0x4EAB5F5: ??? (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28)
  by 0x40101D9: call_init.part.0 (dl-init.c:78)
  by 0x40102C2: call_init (dl-init.c:36)
  by 0x40102C2: _dl_init (dl-init.c:126)
  by 0x4001299: ??? (in /lib/x86_64-linux-gnu/ld-2.19.so)
```

but that doesn't look like it's related to my changes.

**Comment by fxcoudert** [comment]
Upstream report to Gemmi is here: https://github.com/project-gemmi/gemmi/issues/44

File: tests/formats/cif.cpp
Comment (fxcoudert) [reply tests/formats/cif.cpp:98]: Yes, I tested them manually, not sure if they make much sense as unit tests (if the reading is fine, and the unit cell tests are fine, then there should be no problem).

Comment (Luthaf) [reply tests/formats/cif.cpp:98]: Ok, I might remove them from tests-data to prevent confusion then =)

**Comment by fxcoudert** [comment]
And I have now provided Gemmi with a pull request to fix the bug, so hopefully we'll be good to go once they integrate it.

**Comment by Luthaf** [comment]
If needed, we can also merge the PR in https://github.com/chemfiles/gemmi, although I would rather fix this upstream =)

**Comment by fxcoudert** [comment]
Gemmi is fixed, I'll let you decide if you want to wait for a release tag or if you want to include it as is.

**Comment by Luthaf** [comment]
I've updated the included version, I'll do a final round of review if all tests pass

Could you add the new CIF support to the CHANGELOG and yourself to AUHORS?

**Comment by fxcoudert** [comment]
> Could you add the new CIF support to the CHANGELOG and yourself to AUHORS?

Done.

**Review by Luthaf — APPROVED** [review]
Looks good, I'll merge once Travis is happy!

**Pull request merged**
